pkg/applicationserver/io/packages/redis/pagination.go

```diff
@@ -0,0 +1,76 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redis
+
+import (
+	"context"
+
+	"github.com/go-redis/redis"
+	"go.thethings.network/lorawan-stack/pkg/rpcmetadata"
+)
+
+type paginationOptionsKeyType struct{}
+
+var paginationOptionsKey paginationOptionsKeyType
+
+type paginationOptions struct {
+	limit  int64
+	offset int64
+	total  *int64
+}
+
+// WithPagination instructs the store to paginate the results, and set the total
+// number of results into total.
+func WithPagination(ctx context.Context, limit, page int64, total *int64) context.Context {
```

Comment:
IMO this should be `redis.NewContextWithPagination`
We already have e.g. `rights.NewContextWithCache` or `log.NewContextWithField`